src/lib.rs

```diff
@@ -44,7 +44,7 @@ pub fn copy_over<T: Copy>(slice: &mut [T], src_idx: usize, dest_idx: usize, len:
     let ptr = slice.as_mut_ptr();
 
     unsafe {
-        ptr::copy(ptr.add(src_idx), ptr.add(dest_idx), len);
+        ptr::copy(ptr.offset(src_idx as isize), ptr.offset(dest_idx as isize), len);
```

Comment:
Should we check that this cast doesn't overflow? We already bounds-check these indices, and while the docs say that overflowing `isize` is a non-issue for allocations issued by the stdlib, they also point out that it _could_ happen with manually allocated memory on systems with PAE.

I'm also aware that `.add()` also just naively casts to `isize` but I'm wondering if we want to be overly paranoid here and check all preconditions.

Reply:
This isn't an issue. Documentation for [`<[_]>::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) outright says "The total size of the slice must be no larger than isize::MAX bytes in memory." Due to this, it's not possible to create a valid non-ZST slice that takes more than `isize::MAX` bytes in memory.

For ZST types, it is possible to create a slice where `.len() > isize::MAX`. That said, while the cast to `isize` can make the value negative, because `.offset` is essentially a no-op for ZST, providing a wrong value to `.offset` is non-issue, after all, `-50 * 0 == 0`, so the pointer gets moved by 0 bytes, which is still in-bounds for ZST object.

Reply:
Reported the documentation issue for `CStr::from_ptr` in https://github.com/rust-lang/rust/issues/63590.

Reply:
Fair enough.